### PR TITLE
Revision 16

### PR DESCRIPTION
**Pull request by jonathanudd**
* Remove chain spec file override for Turing relay chain
* Fix chain spec url for Tinkernet
* Add Bifrost handling
* [Add GitHub workflow for validating source branch](https://github.com/dwellir-public/polkadot-operator/pull/32/commits/bd157bb8aceef5e3939d23aa6d86b23e574b61f3)
* [Merge pull request](https://github.com/dwellir-public/polkadot-operator/pull/32/commits/17132163f3ebda853a91f0c93c487fb75d96c9ff) https://github.com/dwellir-public/polkadot-operator/pull/33 [from Maharacha/sessionkeys](https://github.com/dwellir-public/polkadot-operator/pull/32/commits/17132163f3ebda853a91f0c93c487fb75d96c9ff)
* [Merge pull request](https://github.com/dwellir-public/polkadot-operator/pull/32/commits/0b65202de8a19294465fd4b2f2c8677f526e6f0f) https://github.com/dwellir-public/polkadot-operator/pull/35 [from Maharacha/rpc-interface](https://github.com/dwellir-public/polkadot-operator/pull/32/commits/0b65202de8a19294465fd4b2f2c8677f526e6f0f)

### CODE REVIEW

**Comment by gflow33** [comment]
Some background, bifrost has a polkadot and kusama parachain, with different chain specs. Prior to deploying the bifrost polkadot parachain, we only had to worry about the kusama parachain. The changes made are to handle both cases.

File: src/service_args.py [hunk 1]
```diff
@@ -254,3 +244,16 @@ def __asset_hub_rococo(self):
         chain_json_path = f'{utils.CHAIN_SPEC_PATH}/{self.chain_name}.json'
         utils.download_chain_spec(chain_json_url, f'{self.chain_name}.json')
         self.__replace_chain_name(chain_json_path, 0)
+
+    def __bifrost(self):
+        if self.chain_name == 'bifrost':
+            chain_json_url = 'https://raw.githubusercontent.com/bifrost-finance/bifrost/develop/node/service/res/bifrost-kusama.json'
+        elif self.chain_name == 'bifrost-polkadot':
```
Comment (jakobilobi) [inline src/service_args.py:251]: @gflow33 

What do you think about renaming `'bifrost'` -> `'bifrost-kusama'`? To keep it similar to the other one. Would also better reflect the actual spec file name.

@jonathanudd Do you agree?

Comment (gflow33) [reply src/service_args.py:251]: I think the assumption I made is that the chain_name is gotten from the `--chain` flag in the service-args and with the current bifrost kusama parachain deployment it is `bifrost`


`
      "value": "--name=\"\\U0001F6E1 DWELLIR BIFROST KUSAMA RPC \\U0001F6E1\" --chain=bifrost --pruning=archive --ws-external --ws-port=9966 --ws-max-connections=3000 --execution=wasm --wasm-execution=compiled --rpc-external --rpc-port=9933 --rpc-cors=all --rpc-methods=Safe --prometheus-external --runtime-cache-size=10 --max-runtime-instances=40 --relay-chain-rpc-urls=ws://192.168.211.224:9933"
`
Except I have a wrong understanding of how the code works?

Me and @jonathanudd went through this when I was deploying the polkadot version and it was named `bifrost-kusama` initially, although given time constraints these changes were used in a local charm instead.

Comment (jakobilobi) [reply src/service_args.py:251]: Yes, I know that's where `chain_name` comes from and no you're not missing how it works.

My proposal is that `bifrost` changes to `bifrost-kusama` in both places, here and for the service args, to make it less confusing when using either one of the Bifrost chains. Would make us have

- `bifrost-kusama`
- `bifrost-polkadot`

and everyone would know what's what :)

Comment (gflow33) [reply src/service_args.py]: Yes I agree with you, makes it more understandable at first glance

Comment (jakobilobi) [reply src/service_args.py]: This means that after the next revision is released, we'll have to update the service args for the bifrost nodes when refreshing them. Let's just make sure to remember :)

Comment (jakobilobi) [reply src/service_args.py]: It doesn't? E.g. the client recognizes `bifrost` as `bifrost-kusama` -> no spec file is needed?

It so we should probably not have it in `service_args.py` at all :thinking:

Comment (gflow33) [reply src/service_args.py]: IIRC I was supposed to deploy a bifrost polkadot node, and then the charm only had support for the kusama node. Then we had to modify the polkadot charm to cover both instances. The kusama node has always been there and I am not really sure of the behaviour. I remember us testing out with the `--chain` and `bifrost` seemed to be the polkadot node but I am not entirely sure of this.

**Comment by erik78se** [comment]
We should probably/possibly use the juju feature "resource" for spec-files which allows supplying a file which is then available to tje charm. It can also be upgraded separately.

**Comment by jakobilobi** [comment]
> We should probably/possibly use the juju feature "resource" for spec-files which allows supplying a file which is then available to tje charm. It can also be upgraded separately.

That would mean we'd have to download the file manually before supplying it to the charm?

File: src/service_args.py (hunk 1 above)
Comment (gflow33) [reply src/service_args.py]: Yes I will remove it

**Comment by erik78se** [comment]
> That would mean we'd have to download the file manually before supplying it to the charm?

Yes and no. A resource can be released together with a charm and be upgraded individually. Eg, we can have a default resource for the default deploy and if it differs it can be supplied.

Further more, we can/could have a set of chain specs built in to the charm, which will be used primarily, whereas, a supplied resource (specc file) only would be used *if* supplied.

Yet another approach would be that the resource is a zip archive of all our supported chains and a config parameter that controls which will be used. 

There might be even more good approaches which gives determinism in what ends up in a deployment.

**Comment by jakobilobi** [comment]
> Further more, we can/could have a set of chain specs built in to the charm, which will be used primarily, whereas, a supplied resource (specc file) only would be used _if_ supplied.

To me, this seems like the only alternative which might be an improvement. The other ones sound like they would mean we have to do more work than what we currently have to do with Polkadot charm deployments. This alternative is basically what we have today (set of chain specs built into the charm, as hardcoded download URL:s), but with a more streamlined way of overriding the specfile if needed (today we can override it by manually adding a spec file and then point to it in the service args).

File: src/service_args.py [hunk 2]
```diff
@@ -123,8 +123,6 @@ def __customize_service_args(self):
             self.__origintrail()
         elif self.chain_name == 'asset-hub-rococo':
             self.__asset_hub_rococo()
-        elif self.chain_name in ['bifrost-kusama', 'bifrost-polkadot']:
-            self.__bifrost()
```
Comment (jakobilobi) [inline src/service_args.py]: Just checking, did you test deploying the charm with these chains after removing this? @gflow33

Comment (gflow33) [reply src/service_args.py]: Yes, I tested it locally and it worked

**Review by jakobilobi — APPROVED** [review]

**Pull request merged**